src/nvim/mbyte.c

```diff
@@ -528,6 +528,76 @@ int utf_ptr2cells(const char *p)
   return 1;
 }
 
+/// Convert a UTF-8 byte sequence to a character number.
+/// Doesn't handle ascii! only multibyte and illegal sequences.
+///
+/// @param[in]  p      String to convert.
+/// @param[in]  len    Length of the character in bytes, 0 or 1 if illegal.
+///
+/// @return Unicode codepoint. A negative value When the sequence is illegal.
+int32_t utf_ptr2CharInfo_impl(uint8_t const *p, uintptr_t const len)
+  FUNC_ATTR_PURE FUNC_ATTR_NONNULL_ALL FUNC_ATTR_WARN_UNUSED_RESULT
+{
+// uint8_t is a reminder for clang to use smaller cmp
+#define CHECK if (EXPECT((uint8_t)(cur & 0xC0U) != 0x80U, false)) goto errorret
+
+  static uint32_t const corrections[] = {
+    (1U << 31) + 0x80U,  // invalid - set invalid bits (safe to add as first 2 bytes
+    (1U << 31) + 0x80U,  // won't affect highest bit in normal ret). 0x80U for errorret
+    -(0x80U + (0xC0U << 6)),  // multibyte - subtract added UTF8 bits (1..10xxx and 10xxx)
+    -(0x80U + (0x80U << 6) + (0xE0U << 12)),
+    -(0x80U + (0x80U << 6) + (0x80U << 12) + (0xF0U << 18)),
+    -(0x80U + (0x80U << 6) + (0x80U << 12) + (0x80U << 18) + (0xF8U << 24)),
+    -(0x80U + (0x80U << 6) + (0x80U << 12) + (0x80U << 18) + (0x80U << 24) + (0xF4U << 30)),
+  };
+
+  // len is 0-6, but declared uintptr_t to avoid zeroing out upper bits
+  uint32_t const corr = corrections[len];
+  uint8_t cur;
+
+  // reading second byte unconditionally, safe for invalid
+  // as it cannot be the last byte, not safe for ascii
+  uint32_t code_point = ((uint32_t)p[0] << 6) + (cur = p[1]);
+  CHECK;
+  if ((uint32_t)len < 3) {
+    goto ret;  // len == 0, 1, 2
+  }
+
+  code_point = (code_point << 6) + (cur = p[2]);
+  CHECK;
+  if ((uint32_t)len == 3) {
+    goto ret;
+  }
+
+  code_point = (code_point << 6) + (cur = p[3]);
+  CHECK;
+  if ((uint32_t)len == 4) {
+    goto ret;
+  }
+
+  code_point = (code_point << 6) + (cur = p[4]);
+  CHECK;
+  if ((uint32_t)len == 5) {
+    goto ret;
+  }
+
+  code_point = (code_point << 6) + (cur = p[5]);
+  if ((uint8_t)(cur & 0xC0U) != 0x80U) {
+errorret:;
+    // 8th bit of corr is always 1, int32_t guerantees
+    // 2's complement. Value is negative
+    int32_t const result = (int32_t)(corr << 24);
+    ASSUME(result < 0);
+    return result;
```

Comment:
If I understand correctly, when this function returns a negative value, the actual value doesn't really matter, so I think it's less confusing if it just returns `-1`.
```suggestion
errorret:;
    return -1;
```

Reply:
If not then at least fix the typo "guerantee"
```suggestion
errorret:;
    // 8th bit of corr is always 1, int32_t guarantees
    // 2's complement. Value is negative
    int32_t const result = (int32_t)(corr << 24);
    ASSUME(result < 0);
    return result;
```